File: 2015/day24/day24.py
#!/usr/bin/env python3
# [Day 24: It Hangs in the Balance](https://adventofcode.com/2015/day/24)

import itertools
import sys
from functools import reduce
from operator import mul
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
self_tests = "-T" in sys.argv
if self_tests:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()

packages = list(map(int, data.splitlines()))


def solve(ngroups):
    weight = sum(packages) // ngroups
    for n in range(1, len(packages) + 1):
        g = [x for x in itertools.combinations(packages, n) if sum(x) == weight]
        if g:
            return min(reduce(mul, p) for p in g)


print(solve(3))
print(solve(4))

File: 2015/day24/day24.rs
//! [Day 24: It Hangs in the Balance](https://adventofcode.com/2015/day/24)

use itertools::Itertools;

struct Puzzle {
    packages: Vec<u64>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            packages: data.lines().filter_map(|s| s.parse().ok()).collect(),
        }
    }

    fn solve(&self, ngroups: u64) -> u64 {
        let weight = self.packages.iter().sum::<u64>() / ngroups;

        for k in 0..self.packages.len() {
            let g = self
                .packages
                .iter()
                .combinations(k)
                .filter(|x| x.iter().copied().sum::<u64>() == weight);

            if let Some(m) = g.map(|p| p.iter().copied().product::<u64>()).min() {
                return m;
            }
        }

        0
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        self.solve(3)
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        self.solve(4)
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 99);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 44);
    }
}
